include/SH3/system/glvertarray.hpp

```diff
@@ -0,0 +1,291 @@
+     *
+     *  @tparam numbuffers  Number of buffers.
+     *  @param attrib       Attribute we want to map data input in our shader (using location = x).
+     *  @param type         Data Type of data passed in.
+     *  @param size         The size of each component per attribute.
+     *  @param stride       Byte offset between vertex attributes.
+     *  @param offset       Offset to attribute in buffer. Each attrib is then located at offset + stride.
+     */
+    template<std::size_t numbuffers>
+    void mutablevao<numbuffers>::SetDataLocation(GLuint attrib, DataType type, GLint size, GLsizei stride, GLint offset)
+    {
+        Bind();
+        glEnableVertexAttribArray(attrib);
+        ASSERT_MSG(attrib < numbuffers, "VBO index (attribute) out of range");
+        buffers[attrib].Bind();
+        glVertexAttribPointer(attrib, size, static_cast<GLenum>(type), GL_FALSE, stride, buffer_offset(offset));
+    }
+
+    /**
+     *  A VAO with indexed vertices.
+     *
+     *  You probably want to use @ref vao.
+     *
+     *  @tparam numbuffers  Number of buffers that should be reserved in @ref buffers.
+     *  @tparam index_slot  The slot number where our index @ref buffer_object resides.
+     */
+    template<std::size_t numbuffers, std::size_t index_slot>
+
+    struct indexedmutablevao : public mutablevao<numbuffers>
+    {
+    protected:
+        /**
+         *  Constructor.
+         *
+         *  @param vertices Reference to our vertices @ref buffer_object.
+         *
+         *  @note This should not be directly instantiated.
+         */
+        indexedmutablevao(const buffer_object& vertices): mutablevao<numbuffers>(mutablevao<numbuffers>::buffers[index_slot], numbuffers, vertices) {}
+    };
+
+    /**
+     *  Declares parent type for @ref vao.
+     *
+     *  Either @ref indexedmutablevao or @ref mutablevao (unindexed).
+     *
+     *  @tparam Attribute An attribute enum.
+     *
+     *  @p Attribute should have a value @c MAX to indicate the number of attributes (slots).
+     *  If @p Attribute has a value @c INDEX_SLOT, then the @ref type is an @ref indexedmutablevao.
+     *
+     *  @note This class cannot be instantiated!
+     */
+    template<typename Attribute>
+    struct vaoparent final
+    {
+    private:
+        template<typename T>
+        static constexpr indexedmutablevao<T::MAX, T::INDEX_SLOT> test(decltype(T::INDEX_SLOT)*);
+
+        template<typename T>
+        static constexpr mutablevao<T::MAX> test(...);
+
+    public:
+        using type = decltype(test<Attribute>(0)); /**< Type of vao this is (either @ref mutablevao or @ref indexedmutablevao) */
+
+    private:
+        vaoparent() = delete;
+    };
+
+    /**
+     *  Our VAO object. This handles all @ref buffer_objects needed to draw data that has been
+     *  sent to the graphics card.
+     *
+     *  @tparam Attribute   The attribute enum that contains all slots required by this VAO. This
```

Comment:
This is a very long parameter description. Do something like the `Attribute` doc for `vaoparent`.